OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java

```diff
@@ -572,139 +443,177 @@ private static OSEmailSubscriptionState getLastEmailSubscriptionState(Context co
-         if (sender_id != null && sender_id.length() > 4)
-            sender_id = sender_id.substring(4);
+      appId = newAppId;
 
-         OneSignal.init(context, sender_id, bundle.getString("onesignal_app_id"), mInitBuilder.mNotificationOpenedHandler, mInitBuilder.mNotificationReceivedHandler);
-      } catch (Throwable t) {
-         t.printStackTrace();
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppId(id) finished, checking if appContext has been set before proceeding...");
+      if (appContext == null) {
+         OneSignal.onesignalLog(LOG_LEVEL.WARN, "appId set, but please call setAppContext(appContext) with Application context to complete OneSignal init!");
+         return;
+      }
+
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppId(id) successful and appContext is set, continuing OneSignal init...");
+      init();
+   }
+
+   /**
+    * 1/2 steps in OneSignal init, relying on setAppId (usage order does not matter)
+    * Sets the global shared ApplicationContext for OneSignal
+    * This is should be set from all OneSignal entry points
+    *   - BroadcastReceivers, Services, and Activities
+    * @param context - Context used by the Application of the app
+    */
+   public static void setAppContext(@NonNull Context context) {
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppContext(context) called!");
+
+      if (context == null) {
+         Log(LOG_LEVEL.WARN, "context is null, ignoring!");
+         return;
       }
+
+      boolean wasAppContextNull = (appContext == null);
+      appContext = context.getApplicationContext();
+      setupActivityLifecycleListener(wasAppContextNull);
+      setupPrivacyConsent(appContext);
+
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppContext(context) finished, checking if appId has been set before proceeding...");
+      if (appId == null) {
+         OneSignal.onesignalLog(LOG_LEVEL.WARN, "appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!");
+         return;
+      }
+
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppContext(context) successful and appId is set, continuing OneSignal init...");
+      init();
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId) {
-      init(context, googleProjectNumber, oneSignalAppId, null, null);
+   public static void setNotificationWillShowInForegroundHandler(NotificationWillShowInForegroundHandler callback) {
+      notificationWillShowInForegroundHandler = callback;
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId, NotificationOpenedHandler notificationOpenedHandler) {
-      init(context, googleProjectNumber, oneSignalAppId, notificationOpenedHandler, null);
+   public static void setNotificationOpenedHandler(NotificationOpenedHandler callback) {
+      notificationOpenedHandler = callback;
+      if (appContext != null)
```

Comment:
This seems like a hack as we setting the same value if it exists. We just process the the list of unfired notifications here if there is any and init is done

Reply:
I have a better idea, Ill see what i can do